File: packages/sdk-go/errors_test.go
package polarissdk

import (
	"context"
	"errors"
	"net"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func TestParseAPIErrorWithEnvelope(t *testing.T) {
	body := []byte(`{"error":{"code":"rate_limited","message":"slow down","retryable":true,"request_id":"req_abc"}}`)
	h := http.Header{}
	h.Set("X-Request-Id", "req_header_fallback")
	apiErr := ParseAPIError(429, body, h)
	if apiErr.Code != "rate_limited" {
		t.Fatalf("code = %q", apiErr.Code)
	}
	if !apiErr.Retryable {
		t.Fatal("expected retryable=true")
	}
	if apiErr.Status != 429 {
		t.Fatalf("status = %d", apiErr.Status)
	}
	// Envelope's request_id wins over the header.
	if apiErr.RequestID != "req_abc" {
		t.Fatalf("request_id = %q", apiErr.RequestID)
	}
	if got := apiErr.Error(); got != "polaris: rate_limited: slow down" {
		t.Fatalf("Error() = %q", got)
	}
}

func TestParseAPIErrorEnvelopeNoRequestID(t *testing.T) {
	// Envelope lacks request_id; the header should fill in.
	body := []byte(`{"error":{"code":"forbidden","message":"nope","retryable":false}}`)
	h := http.Header{}
	h.Set("X-Request-Id", "req_from_header")
	apiErr := ParseAPIError(403, body, h)
	if apiErr.RequestID != "req_from_header" {
		t.Fatalf("request_id = %q", apiErr.RequestID)
	}
	if apiErr.Retryable {
		t.Fatal("expected retryable=false")
	}
}

func TestParseAPIErrorNonJSONBody(t *testing.T) {
	// CF returns an HTML 502 page; the SDK must still produce a typed error.
	body := []byte("<html><body>Bad gateway</body></html>")
	h := http.Header{}
	h.Set("X-Request-Id", "req_502")
	apiErr := ParseAPIError(502, body, h)
	if apiErr.Code != "unknown" {
		t.Fatalf("code = %q", apiErr.Code)
	}
	if !apiErr.Retryable {
		t.Fatal("status >= 500 should be retryable=true on parse failure")
	}
	if apiErr.RequestID != "req_502" {
		t.Fatalf("request_id = %q", apiErr.RequestID)
	}
}

func TestParseAPIErrorEmptyBodyNilHeaders(t *testing.T) {
	apiErr := ParseAPIError(404, nil, nil)
	if apiErr.Code != "unknown" {
		t.Fatalf("code = %q", apiErr.Code)
	}
	if apiErr.Retryable {
		t.Fatal("status < 500 should be retryable=false on parse failure")
	}
	if apiErr.RequestID != "" {
		t.Fatalf("request_id = %q", apiErr.RequestID)
	}
}

func TestDoJSONReturnsAPIErrorOnNon2xx(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/messages/x", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("X-Request-Id", "req_xyz")
		w.WriteHeader(http.StatusTooManyRequests)
		_, _ = w.Write([]byte(`{"error":{"code":"rate_limited","message":"slow down","retryable":true,"request_id":"req_xyz"}}`))
	})
	c, srv := newTestClient(t, mux)
	defer srv.Close()

	err := c.DeleteMessage(context.Background(), "x")
	if err == nil {
		t.Fatal("expected error")
	}
	var apiErr *APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("expected *APIError, got %T: %v", err, err)
	}
	if apiErr.Code != "rate_limited" || apiErr.Status != 429 || !apiErr.Retryable || apiErr.RequestID != "req_xyz" {
		t.Fatalf("apiErr = %+v", apiErr)
	}
	if !IsRetryable(err) {
		t.Fatal("IsRetryable should return true for retryable APIError")
	}
}

func TestDoJSONReturnsAPIErrorOnHTML502(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/messages/get", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/html")
		w.Header().Set("X-Request-Id", "req_502")
		w.WriteHeader(http.StatusBadGateway)
		_, _ = w.Write([]byte("<html>bad gateway</html>"))
	})
	c, srv := newTestClient(t, mux)
	defer srv.Close()

	_, err := c.BulkGetMessages(context.Background(), []string{"a"})
	if err == nil {
		t.Fatal("expected error")
	}
	apiErr, ok := AsAPIError(err)
	if !ok {
		t.Fatalf("expected *APIError, got %T: %v", err, err)
	}
	if apiErr.Code != "unknown" || apiErr.Status != 502 || !apiErr.Retryable {
		t.Fatalf("apiErr = %+v", apiErr)
	}
	if !IsRetryable(err) {
		t.Fatal("502 should be retryable")
	}
}

func TestDoJSONReturnsAPIErrorOnNonRetryable4xx(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/messages/x", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusForbidden)
		_, _ = w.Write([]byte(`{"error":{"code":"scope_violation","message":"from outside scope","retryable":false,"request_id":"req_403"}}`))
	})
	c, srv := newTestClient(t, mux)
	defer srv.Close()

	err := c.DeleteMessage(context.Background(), "x")
	if err == nil {
		t.Fatal("expected error")
	}
	apiErr, ok := AsAPIError(err)
	if !ok {
		t.Fatalf("expected *APIError, got %T: %v", err, err)
	}
	if apiErr.Code != "scope_violation" || apiErr.Retryable {
		t.Fatalf("apiErr = %+v", apiErr)
	}
	if IsRetryable(err) {
		t.Fatal("scope_violation must not be retryable")
	}
}

func TestDoReturnsNetworkErrorUnwrapped(t *testing.T) {
	// Point the client at a closed server so the request errors at the
	// transport layer. The error must not be an *APIError so callers can
	// still use errors.Is for transport-level conditions.
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
	srv.Close() // close immediately
	c := NewClient(srv.URL)
	c.KeyID = "01H0000000000000000000000K"
	c.KeySecret = []byte("test-secret-key")
	c.HTTPClient = &http.Client{Timeout: 100 * time.Millisecond}

	err := c.DeleteMessage(context.Background(), "x")
	if err == nil {
		t.Fatal("expected transport error")
	}
	if _, ok := AsAPIError(err); ok {
		t.Fatalf("transport error must not be *APIError: %v", err)
	}
}

func TestIsRetryableTimeout(t *testing.T) {
	// A simulated net.Error with Timeout()=true must be flagged retryable.
	err := &fakeNetErr{timeout: true}
	if !IsRetryable(err) {
		t.Fatal("IsRetryable should be true for timeout net.Error")
	}
}

func TestIsRetryableNonNetNonAPI(t *testing.T) {
	if IsRetryable(errors.New("oops")) {
		t.Fatal("unknown error must not be retryable")
	}
	if IsRetryable(nil) {
		t.Fatal("nil error must not be retryable")
	}
}

func TestAPIErrorErrorEmptyMessage(t *testing.T) {
	apiErr := &APIError{Code: "unknown", Status: 500}
	if got := apiErr.Error(); got != "polaris: unknown" {
		t.Fatalf("Error() = %q", got)
	}
}

// fakeNetErr satisfies net.Error so we can exercise IsRetryable's timeout
// branch deterministically. The standard-library net.OpError types depend on
// system-level conditions that are hard to reproduce in a unit test.
type fakeNetErr struct {
	timeout bool
}

func (f *fakeNetErr) Error() string   { return "fake net error" }
func (f *fakeNetErr) Timeout() bool   { return f.timeout }
func (f *fakeNetErr) Temporary() bool { return false } //nolint:staticcheck

// Compile-time assertion that fakeNetErr satisfies net.Error.
var _ net.Error = (*fakeNetErr)(nil)

File: packages/sdk-node/test/errors.test.ts
import { describe, expect, it } from 'vitest';
import { Polaris } from '../src/index.js';
import {
  PolarisError,
  isPolarisError,
  isRetryable,
  parsePolarisError,
} from '../src/errors.js';

function jsonResponse(status: number, body: unknown, headers: Record<string, string> = {}): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { 'content-type': 'application/json', ...headers },
  });
}

function textResponse(status: number, body: string, headers: Record<string, string> = {}): Response {
  return new Response(body, {
    status,
    headers: { 'content-type': 'text/html', ...headers },
  });
}

describe('parsePolarisError', () => {
  it('parses the standard envelope', () => {
    const err = parsePolarisError(
      429,
      {
        error: {
          code: 'rate_limited',
          message: 'slow down',
          retryable: true,
          request_id: 'req_abc',
        },
      },
      null,
    );
    expect(err).toBeInstanceOf(PolarisError);
    expect(err.code).toBe('rate_limited');
    expect(err.message).toBe('slow down');
    expect(err.status).toBe(429);
    expect(err.retryable).toBe(true);
    expect(err.requestId).toBe('req_abc');
    expect(err.name).toBe('PolarisError');
  });

  it('falls back to the X-Request-Id header when the envelope omits request_id', () => {
    const err = parsePolarisError(
      403,
      { error: { code: 'forbidden', message: 'no', retryable: false } },
      new Headers({ 'x-request-id': 'req_header' }),
    );
    expect(err.requestId).toBe('req_header');
    expect(err.retryable).toBe(false);
  });

  it('handles non-JSON 502 bodies with code "unknown" and retryable=true', () => {
    const err = parsePolarisError(
      502,
      '<html><body>Bad gateway</body></html>',
      { 'X-Request-Id': 'req_502' },
    );
    expect(err.code).toBe('unknown');
    expect(err.status).toBe(502);
    expect(err.retryable).toBe(true);
    expect(err.requestId).toBe('req_502');
  });

  it('handles non-JSON 404 bodies with retryable=false', () => {
    const err = parsePolarisError(404, 'not found', null);
    expect(err.code).toBe('unknown');
    expect(err.retryable).toBe(false);
    expect(err.requestId).toBeNull();
  });

  it('parses a stringified envelope body', () => {
    const body = JSON.stringify({
      error: { code: 'scope_violation', message: 'nope', retryable: false, request_id: 'req_403' },
    });
    const err = parsePolarisError(403, body, null);
    expect(err.code).toBe('scope_violation');
    expect(err.requestId).toBe('req_403');
  });

  it('treats an envelope without `code` as a parse failure', () => {
    const err = parsePolarisError(400, { error: { message: 'no code field' } }, null);
    expect(err.code).toBe('unknown');
    expect(err.retryable).toBe(false);
  });
});

describe('isPolarisError', () => {
  it('returns true for PolarisError instances', () => {
    expect(isPolarisError(new PolarisError('x', 'y', 500, true, null))).toBe(true);
  });
  it('returns false for plain Errors', () => {
    expect(isPolarisError(new Error('x'))).toBe(false);
  });
  it('returns false for non-errors', () => {
    expect(isPolarisError(null)).toBe(false);
    expect(isPolarisError({ code: 'rate_limited' })).toBe(false);
  });
});

describe('isRetryable', () => {
  it('mirrors PolarisError.retryable', () => {
    expect(isRetryable(new PolarisError('x', 'rate_limited', 429, true, null))).toBe(true);
    expect(isRetryable(new PolarisError('x', 'scope_violation', 403, false, null))).toBe(false);
  });
  it('returns false for unknown errors', () => {
    expect(isRetryable(new Error('boom'))).toBe(false);
    expect(isRetryable(null)).toBe(false);
  });
});

describe('Polaris client throws PolarisError on non-2xx', () => {
  it('throws with the parsed envelope on a 429', async () => {
    const stub = (async () =>
      jsonResponse(
        429,
        {
          error: {
            code: 'rate_limited',
            message: 'slow down',
            retryable: true,
            request_id: 'req_abc',
          },
        },
        { 'x-request-id': 'req_abc' },
      )) as unknown as typeof fetch;
    const client = new Polaris({ baseUrl: 'https://api.example.test', fetch: stub });

    try {
      await client.listMessages();
      throw new Error('expected throw');
    } catch (e) {
      expect(isPolarisError(e)).toBe(true);
      if (isPolarisError(e)) {
        expect(e.code).toBe('rate_limited');
        expect(e.status).toBe(429);
        expect(e.retryable).toBe(true);
        expect(e.requestId).toBe('req_abc');
        expect(isRetryable(e)).toBe(true);
      }
    }
  });

  it('throws PolarisError("unknown") on a non-JSON 502', async () => {
    const stub = (async () =>
      textResponse(502, '<html>bad gateway</html>', {
        'x-request-id': 'req_502',
      })) as unknown as typeof fetch;
    const client = new Polaris({ baseUrl: 'https://api.example.test', fetch: stub });

    try {
      await client.listMessages();
      throw new Error('expected throw');
    } catch (e) {
      expect(isPolarisError(e)).toBe(true);
      if (isPolarisError(e)) {
        expect(e.code).toBe('unknown');
        expect(e.status).toBe(502);
        expect(e.retryable).toBe(true);
        expect(e.requestId).toBe('req_502');
      }
    }
  });

  it('throws PolarisError on a 403 with retryable=false', async () => {
    const stub = (async () =>
      jsonResponse(403, {
        error: {
          code: 'scope_violation',
          message: 'from outside scope',
          retryable: false,
          request_id: 'req_403',
        },
      })) as unknown as typeof fetch;
    const client = new Polaris({ baseUrl: 'https://api.example.test', fetch: stub });

    try {
      await client.sendMessage({ from: 'a@b', to: ['c@d'], subject: '', text: '' } as never);
      throw new Error('expected throw');
    } catch (e) {
      expect(isPolarisError(e)).toBe(true);
      if (isPolarisError(e)) {
        expect(e.code).toBe('scope_violation');
        expect(e.retryable).toBe(false);
        expect(isRetryable(e)).toBe(false);
      }
    }
  });

  it('returns the parsed body on 2xx without throwing', async () => {
    const stub = (async () =>
      jsonResponse(200, { data: [] })) as unknown as typeof fetch;
    const client = new Polaris({ baseUrl: 'https://api.example.test', fetch: stub });
    const res = await client.listMessages();
    expect(res.status).toBe(200);
    expect(res.body).toEqual({ data: [] });
  });
});
